src/node_errors.cc

```diff
@@ -454,6 +454,38 @@ void OnFatalError(const char* location, const char* message) {
   ABORT();
 }
 
+v8::ModifyCodeGenerationFromStringsResult ModifyCodeGenerationFromStrings(
+    v8::Local<v8::Context> context,
+    v8::Local<v8::Value> source,
+    bool is_code_like) {
+  HandleScope scope(context->GetIsolate());
+
+  Environment* env = Environment::GetCurrent(context);
+  Local<Function> maybe_cache_source_map =
+      env->maybe_cache_generated_source_map();
+  {
+    // We do not expect the maybe_cache_generated_source_map to throw any more
+    // exceptions. If it does, just ignore it.
+    errors::TryCatchScope try_catch(env);
+    Local<Value> argv[1] = {source};
+
+    MaybeLocal<Value> maybe_cached = maybe_cache_source_map->Call(
+        context, context->Global(), arraysize(argv), argv);
+    if (maybe_cached.IsEmpty()) {
+      DCHECK(try_catch.HasCaught());
+    }
+  }
+
+  Local<Value> allow_code_gen = context->GetEmbedderData(
+      ContextEmbedderIndex::kAllowCodeGenerationFromStrings);
+  bool codegen_allowed =
+      allow_code_gen->IsUndefined() || allow_code_gen->IsTrue();
+  return {
+      codegen_allowed,
```

Comment:
If no callback is provided, is `codegen_allowed` just always truthy?

Reply:
If the value set in `v8::Context::AllowCodeGenerationFromStrings` is `true`, no host-callbacks are called on compiling dynamic sources. 

The callback is called only when `AllowCodeGenerationFromStrings(false)` is set on the context. Thus if the callback is not provided when `AllowCodeGenerationFromStrings` is set to false, no dynamic sources are allowed.

Reply:
Prior to this, did `v8::Context::AllowCodeGenerationFromStrings` default to `true`?

Any concerns about performance hits we might observe now that we're calling a handler? Perhaps it would be worth adding a [benchmark](https://github.com/nodejs/node/tree/main/benchmark)?

Reply:
@bcoe thank you for your suggestion. I added a new fast-path on if sourcemap is enabled in ModifyCodeGenerationFromStrings and ran a benchmark against the main branch. Now the difference can be slight if the dynamic code is not trivial.

```
                                                confidence improvement accuracy (*)   (**)  (***)
es/eval.js n=1000000 method='without-sourcemap'                -1.01 %       ±1.88% ±2.51% ±3.31%
es/eval.js n=1000000 method='sourcemap'          *     -1.28 %       ±1.04% ±1.39% ±1.82%
```

Reply:
thanks for adding the benchmark 👌

I'm curious, how much more does it slow down if you have the noop closure, without the fast path?

> However, with sourcemap enabled, the difference can be significant. But I suppose this is acceptable since sourcemap comes with costs.

I think this is to be expected. Unrelated to this PR, I'm becoming convinced we should stop performing the logic that loads the original source from disk for providing visual context, as I believe this is one of the major slow downs.

Reply:
The slowdown can be up to ~25% without the fast path with this micro benchmark.

> I think this is to be expected. Unrelated to this PR, I'm becoming convinced we should stop performing the logic that loads the original source from disk for providing visual context, as I believe this is one of the major slow downs.

Yeah, when sourcemap support is not enabled, the visual context is not prepended to the value of the stack property. It's just being prepended when the error is being printed by the runtime like fatal exceptions. While accessing stack property is one of the most commonly used paths, improving the performance of it is always a good motivation. Are you planning to work on it?

Reply:
>  Are you planning to work on it?

I don't think I'll have cycles to work on this in the short term, I'll CC you on the pertinent issue.